File: xform-ts/src/ast.ts
export class Module {
  functions: Record<string, FunctionDef>;
  rules: Record<string, RuleDef[]>;
  vars: Record<string, Expr>;
  namespaces: Record<string, string>;
  imports: Array<[string, string | null]>;
  expr: Expr | null;

  constructor(opts: {
    functions: Record<string, FunctionDef>;
    rules: Record<string, RuleDef[]>;
    vars: Record<string, Expr>;
    namespaces: Record<string, string>;
    imports: Array<[string, string | null]>;
    expr: Expr | null;
  }) {
    this.functions = opts.functions;
    this.rules = opts.rules;
    this.vars = opts.vars;
    this.namespaces = opts.namespaces;
    this.imports = opts.imports;
    this.expr = opts.expr;
  }
}

export type Expr =
  | Literal
  | VarRef
  | IfExpr
  | LetExpr
  | ForExpr
  | MatchExpr
  | FuncCall
  | ApplyExpr
  | UnaryOp
  | BinaryOp
  | PathExpr
  | Constructor
  | TextConstructor
  | CommentConstructor
  | PIConstructor
  | Text
  | Interp;

export class Literal {
  value: any;
  constructor(value: any) {
    this.value = value;
  }
}

export class VarRef {
  name: string;
  constructor(name: string) {
    this.name = name;
  }
}

export class IfExpr {
  cond: Expr;
  then_expr: Expr;
  else_expr: Expr;
  constructor(cond: Expr, thenExpr: Expr, elseExpr: Expr) {
    this.cond = cond;
    this.then_expr = thenExpr;
    this.else_expr = elseExpr;
  }
}

export class LetExpr {
  name: string;
  value: Expr;
  body: Expr;
  constructor(name: string, value: Expr, body: Expr) {
    this.name = name;
    this.value = value;
    this.body = body;
  }
}

export class ForExpr {
  name: string;
  seq: Expr;
  where: Expr | null;
  body: Expr;
  constructor(name: string, seq: Expr, where: Expr | null, body: Expr) {
    this.name = name;
    this.seq = seq;
    this.where = where;
    this.body = body;
  }
}

export class MatchExpr {
  target: Expr;
  cases: Array<[Pattern, Expr]>;
  defaultExpr: Expr | null;
  constructor(target: Expr, cases: Array<[Pattern, Expr]>, defaultExpr: Expr | null) {
    this.target = target;
    this.cases = cases;
    this.defaultExpr = defaultExpr;
  }
}

export class FuncCall {
  name: string;
  args: Expr[];
  namedArgs: Array<[string, Expr]>;
  constructor(name: string, args: Expr[], namedArgs: Array<[string, Expr]> = []) {
    this.name = name;
    this.args = args;
    this.namedArgs = namedArgs;
  }
}

export class ApplyExpr {
  expr: Expr;
  ruleset: string | null;
  constructor(expr: Expr, ruleset: string | null = null) {
    this.expr = expr;
    this.ruleset = ruleset;
  }
}

export class UnaryOp {
  op: string;
  expr: Expr;
  constructor(op: string, expr: Expr) {
    this.op = op;
    this.expr = expr;
  }
}

export class BinaryOp {
  op: string;
  left: Expr;
  right: Expr;
  constructor(op: string, left: Expr, right: Expr) {
    this.op = op;
    this.left = left;
    this.right = right;
  }
}

export class PathExpr {
  start: PathStart;
  steps: PathStep[];
  constructor(start: PathStart, steps: PathStep[]) {
    this.start = start;
    this.steps = steps;
  }
}

export class Constructor {
  name: string;
  attrs: Array<[string, Expr]>;
  contents: Expr[];
  constructor(name: string, attrs: Array<[string, Expr]>, contents: Expr[]) {
    this.name = name;
    this.attrs = attrs;
    this.contents = contents;
  }
}

export class TextConstructor {
  expr: Expr;
  constructor(expr: Expr) {
    this.expr = expr;
  }
}

export class CommentConstructor {
  expr: Expr;
  constructor(expr: Expr) {
    this.expr = expr;
  }
}

export class PIConstructor {
  target: Expr;
  value: Expr;
  constructor(target: Expr, value: Expr) {
    this.target = target;
    this.value = value;
  }
}

export class Text {
  value: string;
  constructor(value: string) {
    this.value = value;
  }
}

export class Interp {
  expr: Expr;
  constructor(expr: Expr) {
    this.expr = expr;
  }
}

export class PathStart {
  kind: string; // context, root, desc, desc_root, var, attr
  name: string | null;
  constructor(kind: string, name: string | null = null) {
    this.kind = kind;
    this.name = name;
  }
}

export class PathStep {
  axis: string; // child, desc, desc_or_self, self, parent, attr
  test: StepTest;
  predicates: Expr[];
  constructor(axis: string, test: StepTest, predicates: Expr[]) {
    this.axis = axis;
    this.test = test;
    this.predicates = predicates;
  }
}

export class StepTest {
  kind: string; // name, wildcard, text, node, comment, pi, document
  name: string | null;
  constructor(kind: string, name: string | null = null) {
    this.kind = kind;
    this.name = name;
  }
}

export type Pattern = WildcardPattern | ElementPattern | TypedPattern | AttributePattern | LiteralPattern;

export class WildcardPattern {}

export class ElementPattern {
  name: string;
  varName: string | null;
  child: Pattern | null;
  attrs: Array<[string, Literal | null]>;
  children: Pattern[];
  constructor(
    name: string,
    varName: string | null = null,
    child: Pattern | null = null,
    attrs: Array<[string, Literal | null]> = [],
    children: Pattern[] = [],
  ) {
    this.name = name;
    this.varName = varName;
    this.child = child;
    this.attrs = attrs;
    this.children = children;
    // Normalize legacy child into children
    if (this.child !== null && this.children.length === 0) {
      this.children = [this.child];
      this.child = null;
    }
  }
}

export class TypedPattern {
  kind: string; // node, text, comment, pi, document
  constructor(kind: string) {
    this.kind = kind;
  }
}

export class AttributePattern {
  name: string;
  value: Literal | null;
  constructor(name: string, value: Literal | null = null) {
    this.name = name;
    this.value = value;
  }
}

export class LiteralPattern {
  value: string;
  constructor(value: string) {
    this.value = value;
  }
}

export class Param {
  name: string;
  type_ref: string | null;
  defaultExpr: Expr | null;
  constructor(name: string, typeRef: string | null = null, defaultExpr: Expr | null = null) {
    this.name = name;
    this.type_ref = typeRef;
    this.defaultExpr = defaultExpr;
  }
}

export class FunctionDef {
  params: Param[];
  body: Expr;
  constructor(params: Param[], body: Expr) {
    this.params = params;
    this.body = body;
  }
}

export class RuleDef {
  pattern: Pattern;
  body: Expr;
  constructor(pattern: Pattern, body: Expr) {
    this.pattern = pattern;
    this.body = body;
  }
}

File: xform-cpp/src/ast.cpp
#include "ast.hpp"
#include "hashmap.hpp"
#include <stdlib.h>
#include <string.h>

void literal_free(LiteralValue *lit) {
    if (!lit) return;
    if (lit->kind == LIT_STR && lit->value.str) {
        free(lit->value.str);
    }
}

void param_free(Param *param) {
    if (!param) return;
    free(param->name);
    free(param->type_ref);
    if (param->default_value) {
        expr_free(param->default_value);
    }
}

void function_def_free(FunctionDef *fd) {
    if (!fd) return;
    for (size_t i = 0; i < fd->param_count; i++) {
        param_free(&fd->params[i]);
    }
    free(fd->params);
    expr_free(fd->body);
    free(fd);
}

void rule_def_free(RuleDef *rd) {
    if (!rd) return;
    pattern_free(rd->pattern);
    expr_free(rd->body);
}

void expr_free(Expr *expr) {
    if (!expr) return;
    
    switch (expr->kind) {
        case EXPR_LITERAL:
            literal_free(&expr->data.literal);
            break;
        case EXPR_VAR_REF:
            free(expr->data.var_ref);
            break;
        case EXPR_IF:
            if (expr->data.if_expr) {
                expr_free(expr->data.if_expr->cond);
                expr_free(expr->data.if_expr->then_expr);
                expr_free(expr->data.if_expr->else_expr);
                free(expr->data.if_expr);
            }
            break;
        case EXPR_LET:
            if (expr->data.let_expr) {
                free(expr->data.let_expr->name);
                expr_free(expr->data.let_expr->value);
                expr_free(expr->data.let_expr->body);
                free(expr->data.let_expr);
            }
            break;
        case EXPR_FOR:
            if (expr->data.for_expr) {
                free(expr->data.for_expr->name);
                expr_free(expr->data.for_expr->seq);
                expr_free(expr->data.for_expr->where_clause);
                expr_free(expr->data.for_expr->body);
                free(expr->data.for_expr);
            }
            break;
        case EXPR_MATCH:
            if (expr->data.match_expr) {
                expr_free(expr->data.match_expr->target);
                for (size_t i = 0; i < expr->data.match_expr->case_count; i++) {
                    pattern_free(expr->data.match_expr->patterns[i]);
                    expr_free(expr->data.match_expr->exprs[i]);
                }
                free(expr->data.match_expr->patterns);
                free(expr->data.match_expr->exprs);
                expr_free(expr->data.match_expr->default_expr);
                free(expr->data.match_expr);
            }
            break;
        case EXPR_FUNC_CALL:
            if (expr->data.func_call) {
                free(expr->data.func_call->name);
                for (size_t i = 0; i < expr->data.func_call->arg_count; i++) {
                    expr_free(expr->data.func_call->args[i]);
                }
                free(expr->data.func_call->args);
                free(expr->data.func_call);
            }
            break;
        case EXPR_UNARY_OP:
            if (expr->data.unary_op) {
                free(expr->data.unary_op->op);
                expr_free(expr->data.unary_op->expr);
                free(expr->data.unary_op);
            }
            break;
        case EXPR_BINARY_OP:
            if (expr->data.binary_op) {
                free(expr->data.binary_op->op);
                expr_free(expr->data.binary_op->left);
                expr_free(expr->data.binary_op->right);
                free(expr->data.binary_op);
            }
            break;
        case EXPR_PATH:
            path_expr_free(expr->data.path);
            break;
        case EXPR_CONSTRUCTOR:
            constructor_free(expr->data.constructor);
            break;
        case EXPR_TEXT_CONSTRUCTOR:
            expr_free(expr->data.text_constructor);
            break;
        case EXPR_CHAR_DATA:
            free(expr->data.char_data);
            break;
        case EXPR_INTERP:
            expr_free(expr->data.interp);
            break;
    }
    free(expr);
}

void pattern_free(Pattern *pat) {
    if (!pat) return;
    switch (pat->kind) {
        case PAT_ELEMENT:
            if (pat->data.element) {
                free(pat->data.element->name);
                free(pat->data.element->var);
                pattern_free(pat->data.element->child);
                free(pat->data.element);
            }
            break;
        case PAT_ATTRIBUTE:
            free(pat->data.attribute);
            break;
        case PAT_TYPED:
            free(pat->data.typed);
            break;
        case PAT_WILDCARD:
            break;
    }
    free(pat);
}

void constructor_free(Constructor *c) {
    if (!c) return;
    free(c->name);
    for (size_t i = 0; i < c->attr_count; i++) {
        free(c->attrs[i].name);
        expr_free(c->attrs[i].expr);
    }
    free(c->attrs);
    for (size_t i = 0; i < c->content_count; i++) {
        expr_free(c->contents[i]);
    }
    free(c->contents);
    free(c);
}

void path_expr_free(PathExpr *pe) {
    if (!pe) return;
    free(pe->start.name);
    for (size_t i = 0; i < pe->step_count; i++) {
        free(pe->steps[i].test.name);
        for (size_t j = 0; j < pe->steps[i].predicate_count; j++) {
            expr_free(pe->steps[i].predicates[j]);
        }
        free(pe->steps[i].predicates);
    }
    free(pe->steps);
    free(pe);
}

void module_free(Module *mod) {
    if (!mod) return;
    hm_free_with_values(mod->functions, (void(*)(void*))function_def_free);
    
    /* Rules hashmap contains arrays of RuleDef* */
    if (mod->rules) {
        size_t count;
        HMEntry *entries = hm_entries(mod->rules, &count);
        for (size_t i = 0; i < count; i++) {
            RuleDef **rules = entries[i].value;
            for (size_t j = 0; rules[j]; j++) {
                rule_def_free(rules[j]);
                free(rules[j]);
            }
            free(rules);
        }
        free(entries);
        hm_free(mod->rules);
    }
    
    hm_free_with_values(mod->vars, (void(*)(void*))expr_free);
    hm_free(mod->namespaces);
    
    for (size_t i = 0; i < mod->import_count; i++) {
        free(mod->imports[i].iri);
        free(mod->imports[i].alias);
    }
    free(mod->imports);
    
    expr_free(mod->expr);
    free(mod);
}

StepTest step_test_named(const char *name) {
    StepTest st;
    st.kind = TEST_NAME;
    st.name = strdup(name);
    return st;
}

StepTest step_test_wildcard(void) {
    StepTest st;
    st.kind = TEST_WILDCARD;
    st.name = NULL;
    return st;
}

StepTest step_test_text(void) {
    StepTest st;
    st.kind = TEST_TEXT;
    st.name = NULL;
    return st;
}

StepTest step_test_node(void) {
    StepTest st;
    st.kind = TEST_NODE;
    st.name = NULL;
    return st;
}

Expr* expr_new(ExprKind kind) {
    Expr *e = calloc(1, sizeof(Expr));
    if (e) e->kind = kind;
    return e;
}

Pattern* pattern_new(PatternKind kind) {
    Pattern *p = calloc(1, sizeof(Pattern));
    if (p) p->kind = kind;
    return p;
}
